Generate Serialization for MediaPlaybackTargetContext.

nishajain61 opened the pull request:
#### f1314fda907d00d4b4f1a88f632f953f5dcd4c32
<pre>
Generate Serialization for MediaPlaybackTargetContext.
<a href="https://bugs.webkit.org/show_bug.cgi?id=268584">https://bugs.webkit.org/show_bug.cgi?id=268584</a>
<a href="https://rdar.apple.com/122137982">rdar://122137982</a>

Reviewed by Chris Dumez.

Generate Serialization for MediaPlaybackTargetContext.
Also removed hand coded NSKU (NSKeyedUnarchiver) encoder and decoder APIs.

* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
(WebCore::MediaPlaybackTargetContext::MediaPlaybackTargetContext):
(WebCore::MediaPlaybackTargetContext::encodingRequiresPlatformData const):
(WebCore::MediaPlaybackTargetContext::serializedOutputContext const): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.mm:
(WebCore::MediaPlaybackTargetContext::encodePlatformData const):
(WebCore::MediaPlaybackTargetContext::hasActiveRoute const):
(WebCore::MediaPlaybackTargetContext::serializeOutputContext): Deleted.
(WebCore::MediaPlaybackTargetContext::deserializeOutputContext): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.serialization.in:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setWirelessPlaybackTarget):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder&lt;WebCore::MediaPlaybackTargetContext&gt;::encodePlatformData): Deleted.
(IPC::ArgumentCoder&lt;WebCore::MediaPlaybackTargetContext&gt;::decodePlatformData): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::Internals::setPlaybackTarget):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::playbackTargetSelected const):

Canonical link: <a href="https://commits.webkit.org/274311@main">https://commits.webkit.org/274311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd4c9ecfe8eb66ddd352d7d03a64343316abb1d1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17577 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41183 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34305 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40955 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20382 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14924 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39221 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14832 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12875 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34476 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42453 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35131 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38692 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13462 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11161 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36897 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15074 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5041 "Built successfully and passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14551 "Build is being retried. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,46 @@ class MediaPlaybackTargetContext {
     {
     }
 
+#if PLATFORM(COCOA)
```
This doesn't make sense, we're already in a cocoa/ folder. We shouldn't need this.

nishajain61 replied:
Fixed

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -89,23 +113,14 @@ class MediaPlaybackTargetContext {
 
 private:
+#if PLATFORM(COCOA)
+    bool m_requirePlatformData;
```
This should have an inline initializer.

nishajain61 replied:
Fixed

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.mm:
```diff
@@ -42,6 +42,22 @@
 }
 
+#if PLATFORM(COCOA)
+std::optional<MediaPlaybackTargetContext::PlatformDataVariant> MediaPlaybackTargetContext::encodePlatformData() const
```
Can we call this `ipcData()` like we do in other places? It is confusing to use "PlatformData" in the name when it may return a NonPlatformData struct.

nishajain61 replied:
Changed the API name

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
```diff
@@ -399,6 +399,8 @@ std::optional<FontPlatformData::Attributes> ArgumentCoder<FontPlatformData::Attr
 #endif
 
 #if ENABLE(WIRELESS_PLAYBACK_TARGET)
+#if !PLATFORM(COCOA)
```
WIRELESS_PLAYBACK_TARGET is only enabled on Cocoa ports so I don't understand the point of these `!PLATFORM(COCOA)` checks.

nishajain61 replied:
Fixed

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,42 @@ class MediaPlaybackTargetContext {
+        RetainPtr<AVOutputContext> outputContext;
+    };
+
+    using PlatformDataVariant = std::variant<struct PlatformData, struct NonPlatformData>;
```
`struct ` is not needed.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,42 @@ class MediaPlaybackTargetContext {
+
+    using PlatformDataVariant = std::variant<struct PlatformData, struct NonPlatformData>;
+
+    MediaPlaybackTargetContext(bool requirePlatformData, Type type, std::optional<PlatformDataVariant>&& platformDataVariant)
```
Does this really need to take in a `std::optional<>`?

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,42 @@ class MediaPlaybackTargetContext {
+            if (platformDataVariant) {
+                WTF::switchOn(*platformDataVariant,
+                    [&] (const PlatformData& d) {
+                        MediaPlaybackTargetContext(d.outputContext.get());
```
I believe you're constructing a new `MediaPlaybackTargetContext ` object here and destroying it right away.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,42 @@ class MediaPlaybackTargetContext {
+                        MediaPlaybackTargetContext(d.outputContext.get());
+                    },
+                    [&] (const NonPlatformData& d) {
+                        MediaPlaybackTargetContext(d.mockDeviceName, d.mockState);
```
ditto.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,42 @@ class MediaPlaybackTargetContext {
+            }
+        }
+
+    WEBCORE_EXPORT std::optional<PlatformDataVariant> encodeIPCData() const;
```
we just use `ipcData()` in other places.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -89,23 +109,13 @@ class MediaPlaybackTargetContext {
-    WEBCORE_EXPORT bool deserializeOutputContext();
-
-    bool encodingRequiresPlatformData() const { return m_type == Type::AVOutputContext || m_type == Type::SerializedAVOutputContext; }
+    bool encodingRequiresPlatformData() const { return m_type == Type::AVOutputContext; }
```
Seems confusing to have both this method and `m_requirePlatformData` with such similar names but being different things.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.mm:
```diff
@@ -42,6 +42,20 @@
 
+std::optional<MediaPlaybackTargetContext::PlatformDataVariant> MediaPlaybackTargetContext::encodeIPCData() const
+{
+    if (MediaPlaybackTargetContext::encodingRequiresPlatformData()) {
```
`MediaPlaybackTargetContext::` is redundant everywhere inside this function implementation.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -44,7 +44,6 @@ enum class MediaPlaybackTargetContextMockState : uint8_t {
 enum class MediaPlaybackTargetContextType : uint8_t {
     None,
     AVOutputContext,
-    SerializedAVOutputContext,
```
It is unclear to me why we're getting rid of this and the associated logic. This should probably be explained in the commit log.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.serialization.in:
```diff
@@ -31,8 +33,22 @@ enum class WebCore::MediaPlaybackTargetContextMockState : uint8_t {
+    RetainPtr<AVOutputContext> outputContext;
+}
+
+[CustomHeader] class WebCore::MediaPlaybackTargetContext {
```
Why does this need `CustomHeader`?

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11660,9 +11660,8 @@ void WebPageProxy::Internals::setPlaybackTarget(PlaybackTargetClientContextIdent
-    ASSERT(context.type() != MediaPlaybackTargetContext::Type::SerializedAVOutputContext);
     if (page.preferences().useGPUProcessForMediaEnabled())
-        context.serializeOutputContext();
+        context.outputContext();
```
What are you trying to achieve here? `outputContext()` looks like a trivial getter and you're not using the value it returns. This looks like a no-op?

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -89,23 +110,15 @@ class MediaPlaybackTargetContext {
 
 private:
+
+    bool m_presentPlatformData { false };
```
What does this mean? The name is very confusing.

cdumez replied:
Maybe you mean `m_hasPlatformData` or `m_isPlatformDataPresent`?

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.serialization.in:
```diff
@@ -31,8 +33,22 @@ enum class WebCore::MediaPlaybackTargetContextMockState : uint8_t {
+class WebCore::MediaPlaybackTargetContext {
+    bool encodingRequiresPlatformData();
+    WebCore::MediaPlaybackTargetContextType type();
+    std::optional<WebCore::MediaPlaybackTargetContext::PlatformDataVariant> ipcData();
```
I think @achristensen07 told me you have to use the full `std::variant<>` type explicitly here instead of the `PlatformDataVariant` alias for that our code generated knows about the type of the IPC data.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,43 @@ class MediaPlaybackTargetContext {
+        MockState mockState;
+    };
+
+    struct PlatformData {
```
I don't think we need this struct, we can use `RetainPtr<AVOutputContext>` directly.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.serialization.in:
```diff
@@ -31,8 +33,22 @@ enum class WebCore::MediaPlaybackTargetContextMockState : uint8_t {
+}
+
+class WebCore::MediaPlaybackTargetContext {
+    bool encodingRequiresPlatformData();
```
Encoding this boolean in the IPC doesn't make sense to me. The recipient can just check if the variant contains platform data or not.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.serialization.in:
```diff
@@ -31,8 +33,22 @@ enum class WebCore::MediaPlaybackTargetContextMockState : uint8_t {
+    WebCore::MediaPlaybackTargetContextMockState mockState;
+}
+
+[Nested] struct WebCore::MediaPlaybackTargetContext::PlatformData {
```
Not needed.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -89,23 +110,15 @@ class MediaPlaybackTargetContext {
 
 private:
+
+    bool m_hasPlatformData { false };
```
Why do we need this boolean, why can't we check if `m_outputContext` is null?

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,43 @@ class MediaPlaybackTargetContext {
+        RetainPtr<AVOutputContext> outputContext;
+    };
+
+    using PlatformDataVariant = std::variant<PlatformData, NonPlatformData>;
```
PlatformDataVariant is not a good name IMO because it may also represent NON platform data. We usually use `IPCData` for this.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,43 @@ class MediaPlaybackTargetContext {
+        , m_type(type)
+        {
+            if (platformDataVariant) {
+                WTF::switchOn(*platformDataVariant,
```
`WTFMove(*platformDataVariant)`

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,43 @@ class MediaPlaybackTargetContext {
+        {
+            if (platformDataVariant) {
+                WTF::switchOn(*platformDataVariant,
+                    [&] (const PlatformData& d) {
```
`RetainPtr<AVOutputContext>&&`

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,43 @@ class MediaPlaybackTargetContext {
+            if (platformDataVariant) {
+                WTF::switchOn(*platformDataVariant,
+                    [&] (const PlatformData& d) {
+                        m_outputContext = d.outputContext.get();
```
`m_outputContext = WTFMove(d.outputContext);`

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,43 @@ class MediaPlaybackTargetContext {
+                        m_outputContext = d.outputContext.get();
+                        m_type = Type::AVOutputContext;
+                    },
+                    [&] (const NonPlatformData& d) {
```
`NonPlatformData&&`

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,43 @@ class MediaPlaybackTargetContext {
+                        m_type = Type::AVOutputContext;
+                    },
+                    [&] (const NonPlatformData& d) {
+                        m_mockDeviceName = d.mockDeviceName;
```
`m_mockDeviceName = WTFMove(d.mockDeviceName);`

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -57,21 +56,43 @@ class MediaPlaybackTargetContext {
+        : m_hasPlatformData(hasPlatformData)
+        , m_type(type)
+        {
+            if (platformDataVariant) {
```
I think we need to fail decoding if this is false, in order to maintain pre-existing behavior. I'd suggest we use `[CreateUsing=fromIPCData]` in the serialization.in. Then you replace this constructor with a static function named `fromIPCData` which returns a `std::optional< MediaPlaybackTargetContext>`. In case of failure (e.g. `platformDataVariant` is `std::nullopt`, have the function return std::nullopt so that decoding fails.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.mm:
```diff
@@ -42,6 +42,31 @@
     m_type = MediaPlaybackTargetContext::Type::AVOutputContext;
 }
 
+MediaPlaybackTargetContext::IPCData MediaPlaybackTargetContext::ipcData() const
```
Another way to write this that we often use:
```
auto MediaPlaybackTargetContext::ipcData() const -> IPCData
```

A little more concise.

cdumez commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.mm:
```diff
@@ -42,6 +42,31 @@
+{
+    if (ipcData) {
+        return WTF::switchOn(WTFMove(*ipcData), [](RetainPtr<AVOutputContext>&& outputContext) {
+            return MediaPlaybackTargetContext { outputContext.get() };
```
should use `WTFMove(outputContext)` to avoid ref-counting churn. The constructor takes an r-value reference.

cdumez reviewed (approved):
LGTM with a couple of nits.

nishajain61 committed:
Generate Serialization for MediaPlaybackTargetContext.
https://bugs.webkit.org/show_bug.cgi?id=268584
rdar://122137982

Reviewed by Chris Dumez.

Generate Serialization for MediaPlaybackTargetContext.
Also removed hand coded NSKU (NSKeyedUnarchiver) encoder and decoder APIs.

* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
(WebCore::MediaPlaybackTargetContext::MediaPlaybackTargetContext):
(WebCore::MediaPlaybackTargetContext::encodingRequiresPlatformData const):
(WebCore::MediaPlaybackTargetContext::serializedOutputContext const): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.mm:
(WebCore::MediaPlaybackTargetContext::encodePlatformData const):
(WebCore::MediaPlaybackTargetContext::hasActiveRoute const):
(WebCore::MediaPlaybackTargetContext::serializeOutputContext): Deleted.
(WebCore::MediaPlaybackTargetContext::deserializeOutputContext): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.serialization.in:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setWirelessPlaybackTarget):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::MediaPlaybackTargetContext>::encodePlatformData): Deleted.
(IPC::ArgumentCoder<WebCore::MediaPlaybackTargetContext>::decodePlatformData): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::Internals::setPlaybackTarget):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::playbackTargetSelected const):

Canonical link: https://commits.webkit.org/274311@main
